zschema/keys.py

```diff
@@ -124,6 +144,67 @@ def __iter_classes(kls):
             Keyable._types_by_name[klass.__name__] = klass
 
 
+    def __init__(self, required=_NO_ARG, desc=_NO_ARG, doc=_NO_ARG, category=_NO_ARG,
+            exclude=_NO_ARG, deprecated=_NO_ARG, ignore=_NO_ARG,
+            examples=_NO_ARG, metadata=_NO_ARG):
+        self.set("required", required)
+        self.set("desc", desc)
+        self.set("doc", doc)
+        self.set("category", category)
+        self.set("exclude", exclude)
+        self.set("examples", examples)
+        self.set("metadata", metadata)
+        self.set("deprecated", deprecated)
+        self.set("ignore", ignore)
+
+        if self.DEPRECATED_TYPE:
+            e = "WARN: %s is deprecated and will be removed in a "\
+                    "future release\n" % self.__class__.__name__
+            sys.stderr.write(e)
+
+    def to_dict(self):
+        retv = {
+            "required":self.required,
+            "doc":self.doc,
+            "type":self.__class__.__name__,
+            "metadata":self.metadata,
+            "examples": self.examples,
+        }
+        return retv
+
+    def set(self, k, v):
+        new_k = "_value_" + k
+        setattr(self, new_k, v)
+
+    @classmethod
+    def set_default(cls, k, v):
+        if v is not _NO_ARG:
+            new_k = k.upper()
```

Comment:
One non-blocking comment here -- I missed this the first pass, and when it got to checking `assertEqual("class docs", ssh.DOC)`, I was rather confused.

Maybe note somewhere prominently that all-caps attributes are reserved, and function as defaults for their non-all-upper equivalents.

I'm not as concerned that `value` and `Value` will have the same default, and you have ruled out the `_value`/`value` problem.